### PR TITLE
OJ-997 - Added new integration test for session endpoint

### PR DESCRIPTION
**Pull request by techaman-m**
### What changed
We have created a happy case scenario to test session endpoint

### Why did it change

This test will act as a QA gateway.

### Issue tracking
 
https://govukverify.atlassian.net/browse/OJ-997
### Environment variables or secrets

STACK_NAME=di-ipv-cri-xxx 
API_GATEWAY_ID_PRIVATE=xxx 
IPV_CORE_STUB_BASIC_AUTH_USER=xxx 
IPV_CORE_STUB_BASIC_AUTH_PASSWORD=xxx 
IPV_CORE_STUB_URL="https://di-ipv-core-stub.london.cloudapps.digital" 

### To run the test

`STACK_NAME=di-ipv-cri-xxx  API_GATEWAY_ID_PRIVATE=xxx  IPV_CORE_STUB_BASIC_AUTH_USER=xxx IPV_CORE_STUB_BASIC_AUTH_PASSWORD=xxx IPV_CORE_STUB_URL="https://di-ipv-core-stub.london.cloudapps.digital"  gradle integration-tests:cucumber`

### CODE REVIEW

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java [hunk 1]
```diff
@@ -1,3 +1,63 @@
...
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private String sessionId;
+
+    @Given("user has the user identity in the form of a signed JWT string")
+    public void user_has_the_user_identity_in_the_form_of_a_signed_jwt_string()
+            throws URISyntaxException, IOException, InterruptedException {
+        int experianRowNumber = 681;
+        String ipvCoreStubURL = IpvCoreStubUtil.getIPVCoreStubURL();
+        String userIdentityJson =
+                IpvCoreStubUtil.getClaimsForUser(ipvCoreStubURL, experianRowNumber);
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java:33]: remove the ipvCoreStubURL method parameter & simply pass the experianRowNumber to the getClaimsForUser method

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java [hunk 2]
```diff
@@ -1,3 +1,63 @@
...
+                HttpRequest.newBuilder()
+                        .uri(
+                                new URIBuilder(IpvCoreStubUtil.getPrivateAPIEndpoint())
+                                        .setPath("/dev/session")
+                                        .build())
+                        .setHeader("Accept", "application/json")
+                        .setHeader("Content-Type", "application/json")
+                        .POST(HttpRequest.BodyPublishers.ofString(sessionRequestBody))
+                        .build();
+        response = IpvCoreStubUtil.sendHttpRequest(request);
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java:50]: add a new method to the IpvCoreStubUtil entitled sendSessionRequest and pass in the sessionRequestBody

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java [hunk 3]
```diff
@@ -1,3 +1,63 @@
...
+        String ipvCoreStubURL = IpvCoreStubUtil.getIPVCoreStubURL();
+        String userIdentityJson =
+                IpvCoreStubUtil.getClaimsForUser(ipvCoreStubURL, experianRowNumber);
+        sessionRequestBody = IpvCoreStubUtil.createRequest(ipvCoreStubURL, userIdentityJson);
+    }
+
+    @When("user sends a request to session end point")
+    public void user_sends_a_request_to_session_end_point()
+            throws URISyntaxException, IOException, InterruptedException {
+        var request =
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/stepDefinitions/SessionEndpointSteps.java:40]: remove lines 40-49

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Objects;
+import java.util.Optional;
+
+public class IpvCoreStubUtil {
+
+    private static final String ADDRESS_CRI_DEV = "address-cri-dev";
+
+    public static String getPrivateAPIEndpoint() {
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java:19]: make this method private

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return getApiEndpoint(
+                "API_GATEWAY_ID_PRIVATE", "Environment variable PRIVATE API endpoint is not set");
+    }
+
+    private static String getApiEndpoint(String apikey, String message) {
+        String apiEndpoint = System.getenv(apikey);
+        Optional.ofNullable(apiEndpoint).orElseThrow(() -> new IllegalArgumentException(message));
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java:26]: remove the second `message` argument from this method.
In the IllegalArgumentException constructor pass in something along the lines of 
String.format("Environment variable %s is not assigned", apikey)

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                                        .addParameter("cri", ADDRESS_CRI_DEV)
+                                        .addParameter(
+                                                "rowNumber", String.valueOf(userDataRowNumber))
+                                        .build())
+                        .GET()
+                        .build();
+        return sendHttpRequest(request).body();
+    }
+
+    public static HttpResponse<String> sendHttpRequest(HttpRequest request)
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java:46]: make this method `private`

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                                        return new PasswordAuthentication(
+                                                basicAuthUser, basicAuthPassword.toCharArray());
+                                    }
+                                })
+                        .build();
+
+        return client.send(request, HttpResponse.BodyHandlers.ofString());
+    }
+
+    public static String getIPVCoreStubURL() {
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java:73]: make this method private

File: integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        return Optional.ofNullable(System.getenv("IPV_CORE_STUB_URL"))
+                .orElseThrow(
+                        () ->
+                                new IllegalArgumentException(
+                                        "Environment variable IPV_CORE_STUB_URL is not set"));
+    }
+
+    public static String createRequest(String baseUrl, String jsonString)
```
Comment (tal-nagra) [inline integration-tests/src/test/java/gov/uk/di/ipv/cri/common/api/util/IpvCoreStubUtil.java:81]: remove the baseUrl parameter & rename this method to sendCreateSessionRequest

**Review by tal-nagra — CHANGES_REQUESTED** [review]

**Review by tal-nagra — DISMISSED** [review]
Any further re-factoring etc can be tackled in a future PR

**Review by idrop — APPROVED** [review]

**Pull request merged**